File: leetcode/js/majority-element.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var majorityElement = function (nums) {
  let result = 0;
  const times = nums.length / 2;
  const count = {};

  for (const num of nums) {
    count[num] = num in count ? count[num] + 1 : 1;

    const prevNum = count[result] || 0;

    if (count[num] >= times && prevNum < count[num]) {
      result = num;
    }
  }

  return result;
};

var majorityElement = function (nums) {
  let result = 0;
  let count = 0;

  for (const num of nums) {
    if (count === 0) {
      result = num;
    }

    count += result === num ? 1 : -1;
  }

  return result;
};

File: leetcode/rust/majority-element.rs
impl Solution {
    pub fn majority_element(nums: Vec<i32>) -> i32 {
        let mut result = 0;
        let mut count = 0;

        for num in nums {
            if count == 0 {
                result = num;
            }

            count += if result == num { 1 } else { -1 };
        }

        result
    }
}
